lens-macro-impl/src/lib.rs

```diff
@@ -0,0 +1,180 @@
+extern crate proc_macro;
+
+use quote::quote;
+use syn::{parse_macro_input, Expr, DeriveInput, Ident, Result, Token};
+use syn::parse::{Parse, ParseStream};
+use proc_macro2::{Span};
+use proc_macro_hack::proc_macro_hack;
+
+#[proc_macro_derive(Lens)]
+pub fn derive_lens(input: proc_macro::TokenStream) -> proc_macro::TokenStream {
+    let input = parse_macro_input!(input as DeriveInput);
+
+    let name = input.ident;
+
+    let fields = if let syn::Data::Struct(syn::DataStruct {
+        fields: syn::Fields::Named(syn::FieldsNamed { ref named, .. }),
+        ..
+    }) = input.data
+    {
+        named
+    } else {
+        unimplemented!();
+    };
+
+    let twizzled_name = if is_camel_case(&name.to_string()) {
+        let temp_name = to_snake_case(&mut name.to_string());
+        Ident::new(&temp_name, Span::call_site())
+    } else {
+        unimplemented!();
```

Comment:
is there some other reasonable path we can take here? like just prepending some prefix and using the name as-is?

Reply:
Or if we want to say that you need to use camelCase, we can make this a `compile_error` with some diagnostic. If we want helpful diagnostics, we can also return a `syn::Error` here, which should be associated with the span of the type.

Reply:
would need some way of telling the user that they need to prefix in this scenario, yes?